### PR TITLE
datapath: Add ctx_redirect_to_proxy_hairpin_ipv6

### LINKED ISSUE #17717: bpf_lxc: ctx_redirect_to_proxy_hairpin decrements IPv4 TTL on IPv6 path
eBPF code in [`bpf_lxc.c` IPv6 ingress to pod path](https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L1263) uses [`ctx_redirect_to_proxy_hairpin()`](https://github.com/cilium/cilium/blob/master/bpf/lib/lxc.h#L59-L92), which calls [`ipv4_l3()`](https://github.com/cilium/cilium/blob/master/bpf/lib/lxc.h#L80), which decrements IPv4 TTL, which will either fail or mess with the fragment offset field of the IPv6 header. This is done for L7 proxy redirection, which can either work or fail depending on the sanity checks on the IPv6 header by the kernel.

### PR DESCRIPTION
**Pull request by jrajahalme**
Add IPv6 version of ctx_redirect_to_proxy_hairpin() and use it in IPv6 path.

Move ctx_redirect_to_proxy_hairpin functions to a new
lib/proxy_hairpin.h so that they can be used from other bpf programs
as well.

Fixes: #17717
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
L7 proxy redirection on IPv6 ingress to a pod is fixed to properly update IPv6 hop limit.
```

### CODE REVIEW

File: bpf/lib/proxy_hairpin.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#include "trace.h"
+#include "csum.h"
+#include "l4.h"
+
+/**
+ * ctx_redirect_to_proxy_hairpin redirects to the proxy by hairpining the
+ * packet out the incoming interface
+ */
+static __always_inline int
+ctx_redirect_to_proxy_hairpin(struct __ctx_buff *ctx, __be16 proxy_port)
```
Comment (qmonnet) [inline bpf/lib/proxy_hairpin.h]: Let's add a suffix to the IPv4 version too?
```suggestion
ctx_redirect_to_proxy_hairpin_ipv4(struct __ctx_buff *ctx, __be16 proxy_port)
```

Comment (jrajahalme) [reply bpf/lib/proxy_hairpin.h]: Done :-)

File: bpf/lib/proxy_hairpin.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+#endif /* ENABLE_IPV4 */
+
+#ifdef ENABLE_IPV6
+/**
+ * ctx_redirect_to_proxy_hairpin_ipv6 redirects to the proxy by hairpining the
+ * packet out the incoming interface
+ */
+static __always_inline int
+ctx_redirect_to_proxy_hairpin_ipv6(struct __ctx_buff *ctx, __be16 proxy_port)
```
Comment (joamaki) [inline bpf/lib/proxy_hairpin.h:74]: I'm not super keen on duplicating this much code when it only differs by two lines. I would suggest instead adding an `const bool is_ipv6` or some such as an argument to the function and then having the function call either ipv4_l3 or ipv6_l3. The compiler should end up generating the exact same code as the argument is a compile-time constant, but we wouldn't end up with an almost identical copy in the C code.

Comment (jrajahalme) [reply bpf/lib/proxy_hairpin.h:74]: These call `ipv4_l3()` or `ipv6_l3()`, which are conditionally defined, which makes this implementation choice a little awkward. I'll try it out :-)

Comment (jrajahalme) [reply bpf/lib/proxy_hairpin.h:74]: I like the call sites better without a random `true`/`false` argument, so came up with this. Do you like this better @joamaki ??
```c
static __always_inline int
ctx_redirect_to_proxy_hairpin(struct __ctx_buff *ctx, __be16 proxy_port, const bool is_ipv6)
{
	union macaddr host_mac = HOST_IFINDEX_MAC;
	union macaddr router_mac = NODE_MAC;
	void *data_end = (void *)(long)ctx->data_end;
	void *data = (void *)(long)ctx->data;
	int ret;

	ctx_store_meta(ctx, CB_PROXY_MAGIC,
		       MARK_MAGIC_TO_PROXY | (proxy_port << 16));
	bpf_barrier(); /* verifier workaround */

	if (is_ipv6) {
#ifdef ENABLE_IPV6
		struct ipv6hdr *ip6;

		if (!revalidate_data(ctx, &data, &data_end, &ip6))
			return DROP_INVALID;

		ret = ipv6_l3(ctx, ETH_HLEN, (__u8 *)&router_mac, (__u8 *)&host_mac, METRIC_EGRESS);
		if (IS_ERR(ret))
			return ret;
#endif
	} else {
#ifdef ENABLE_IPV4
		struct iphdr *ip4;

		if (!revalidate_data(ctx, &data, &data_end, &ip4))
			return DROP_INVALID;

		ret = ipv4_l3(ctx, ETH_HLEN, (__u8 *)&router_mac, (__u8 *)&host_mac, ip4);
		if (IS_ERR(ret))
			return ret;
#endif
	}

	cilium_dbg(ctx, DBG_CAPTURE_PROXY_PRE, proxy_port, 0);

	/* Note that the actual __ctx_buff preparation for submitting the
	 * packet to the proxy will occur in a subsequent program via
	 * ctx_redirect_to_proxy_first().
	 */

	return redirect(HOST_IFINDEX, 0);
}

#ifdef ENABLE_IPV4
static __always_inline int
ctx_redirect_to_proxy_hairpin_ipv4(struct __ctx_buff *ctx, __be16 proxy_port)
{
	return ctx_redirect_to_proxy_hairpin(ctx, proxy_port, false);
}
#endif

#ifdef ENABLE_IPV6
static __always_inline int
ctx_redirect_to_proxy_hairpin_ipv6(struct __ctx_buff *ctx, __be16 proxy_port)
{
	return ctx_redirect_to_proxy_hairpin(ctx, proxy_port, true);
}
#endif
```

Comment (joamaki) [reply bpf/lib/proxy_hairpin.h:74]: Yeah agree on the random true/false being ugly. Wrapping it like this is nice.

**Review by qmonnet — APPROVED** [review]
Looks good to me (provided you address the “unused variable” reports raised by the CI actions). Thanks!

**Review by pchaigno — APPROVED** [review]
Harder to review than it could be since the single commit mixes code movement with the actual fix (:wink:), but the logic is simple and fix looks good to me!

**Comment by jrajahalme** [comment]
Fixed checkpatch warnings on old (but moved) code.

**Review by joestringer — APPROVED** [review]

**Comment by jrajahalme** [comment]
Had to keep using IPv4 header pointer for `revalidata_data()` call instead of `void *` so that accessing TTL (and hop limit for IPv6) is not flagged as invalid by verifier.

**Comment by pchaigno** [comment]
> Had to keep using IPv4 header pointer for `revalidata_data()` call instead of `void *` so that accessing TTL (and hop limit for IPv6) is not flagged as invalid by verifier.

@jrajahalme Not sure if it's intended, but you didn't change anything apart from the variable name: https://github.com/cilium/cilium/compare/1e9ba592a46bc722cd3c167b4b5eeb98092bbde6..0f9d49c1f5e2de5bec45f2a6599fc3044f80c0b1.

**Comment by jrajahalme** [comment]
k8s-1.20 hit flake #17069

**Comment by jrajahalme** [comment]
EKS fail looks similar to #16938

**Comment by pchaigno** [comment]
Multicluster has the usual failures. k8s-1.20-kernel-5.4 failed with known flake https://github.com/cilium/cilium/issues/16852. Reviews are in. Merging.

**Pull request merged**